### ISSUE TITLE
[Change request]: add a "targetGroup" attribute or similar mechanism

### ISSUE THREAD

**Issue by jelmerderonde** [open-education-api/specification#160]
### Organization

SURF

### Project

Project Studentmobiliteit

### Contact Details

jelmer.deronde@surf.nl

### Short description

For the project "Studentmobiliteit" an extra attribute was added to the extension object of the `courses` and `courseOfferings` objects: `targetGroup`. In addition the participating parties decided to also supporting filtering on this attribute.

The `targetGroup` attribute allows clients to request only objects that are meant for a certain application.

### Version

v5

### Usecase

Makes it possible for applications and institutions to only decide to select only certain subsets of education and educationOfferings (and maybe also objects as well).

### Proposed solution

In the pilot, the attribute is currently used as follows:

```json
{
    "targetgroup":
    [{"name": "EWUU",
       "theme": "Natuur"}]
}
```

**Comment by jelmerderonde** [comment]
Zou je een dergelijk mechanisme kunnen gebruiken om aan te geven of iets voor STAP bedoeld is?

**Comment by jelmerderonde** [comment]
@arthurvanalten also had some ideas about using a tag based mechanism.

**Comment by jelmerderonde** [comment]
We should discuss the difference between filtering and authorisation.

**Comment by jelmerderonde** [comment]
@hermanvand de discussie die hier gaat plaatsvinden is ook interessant voor jou.

**Comment by Marcel-Noordzij** [comment]
The discussion to support this attribute should not only be limited to courses and courseOfferings, but also to program level. 
This to allow specification of a targetGroup for a minor or specialisation.

**Comment by mpdejong** [comment]
Targeting / ‘uitbreiding attributen voor specifieke afnemers’:

Dit is een beetje tricky omdat je feitelijk de generieke OOAPI-definitie wilt uitbreiden met afnemer-specifieke attributen.
Om dit toch generiek te kunnen aanbieden zou je kunnen denken aan het mogen uitbreiden van een object met custom attributen. Zo’n custom attribuut moet dan beginnen met een prefix van drie letters en een underscore.
Bijvoorbeeld ‘rio_description’.
Om het voor een afnemer de mogelijkheid te kunnen bieden eigen uitbreidingen te kunnen toevoegen zou dit wellicht met een json-bestandje kunnen waarin de uitbreiding beschreven staat.
Een soort extensie / add-on - mechanisme (die alleen kan worden toegevoegd via de werkgroep?).

Op de OOAPI-docs zou voor afnemers dan de mogelijkheid moeten komen om een add-on te kunnen kiezen (bijvoorbeeld de ‘rio-add-on’) waarmee op alle objecten de custom attributen worden getoond.
Een extensie-definitie zou er bijvoorbeeld als volgt kunnen uitzien:
```
{
  "addon": {
    "name": "rio",
    "extensions": [
      {
        "type": "course",
        "attributes": [
          {
            "name": "rio_attributeX",
            "attributeType": "number",
            "attributeTypeSpec": ">=0",
            "description": "This rio_attrbuteX is needed for scenario X"
          },
          {
            "name": "rio_attributeY",
            "attributeType": "string",
            "attributeTypeSpec": "<= 256 characters",
            "description": "This rio_attrbuteY means Y (or something like that)"
          },
          {
            "name": "rio_attributeZ",
            "attributeType": "enum",
            "attributeTypeSpec": "string",
            "attributeValues": "waarde1, waarde2, waarde3",
            "description": "just a enumeration"
          }
        ]
      },
      {
        "type": "program",
        "attributes": [
          {
            "name": "rio_attributeA",
            "attributeType": "string",
            "description": "iets"
          },
          {
            "name": "rio_attributeB",
            "attributeType": "number",
            "attributeTypeSpec": ">=0",
            "description": "iets anders"
          }
        ]
      }
    ]
  }
}
```

(Voorstel opgevoerd door Arthur van Alten en Mark de Jong.)

**Comment by hermanvand** [comment]
Three remarks regarding this issue

1) I like the idea of an 'add-on'. We could use this to describe extensions that are in use for a particular customer application. But this might not be limited to extensions/add-ons. Other examples of different interpretations of ooapi are

- a customer application might require more attributes, that are not required in the ooapi spec. It would be nice to specify that as well. 
- the add-on could describe the required version of ooapi for the customer application and the specific resources needed. 

Seen this way, the add-on becomes a kind of 'ooapi data profile'. A profile that describes a particular subset of ooapi that is in use by a customer application. That would help communication to institutions for developing their ooapi endpoint. And such a profile could help in validating the data.

2) We already have the ext{} object in place for every resource in ooapi. I think the described add-on extensions should go in this object. 

3) However, the targetGroup attribute should be in the ooapi specification itself, i think. The targetGroup in this issue consist of a name of the group and a theme. I agree that 'theme' could also be a course extension. But targetGroup=[NAME] is different. Because it allows to filter the output for a particular target. A request for /courses?targetGroup=[NAME] results in a list of courses specific for this target. This is a way for institutions to have one OOAPI endpoint and distribute different subsets of data to different customer applications, based on this targetGroup. This attribute transcends a particular customer application.

I think having a targetGroup attribute/filter and having an add-on/data profile description of a customer application are two different things. The targetGroup should be in ooapi V5. The add-on description might need some more work, although it would be very nice to have also for RIO.

**Comment by jelmerderonde** [comment]
Closed by #197
